### PR TITLE
Upgrade numpy to 1.21.4

### PR DESCRIPTION
**Pull request by leopsidom**
### Description

Currently pyodide has numpy @1.17.5 but the newest numpy version is 1.21.4. This PR is to keep the numpy package in pyodide up to date.  Some changes made in this PR include:

1. patches removed: `add-emscripten-cpu.patch` and `rm-duplicate-symbols-link.patch`. These patches got removed because the underlying issue has been fixed in the numpy package.
2. patches added: 
     1. `fix-invalid-asm-instruction.patch` to fix invalid asm instructions
     2. `limit-cpu-baseline.patch` this is to limit cpu baseline to `AVX` so it doesn't try to build SIMD feature AVX512 that emscripten doesn't support;
     3. `not-build-lapack-lite-as-64-bit.patch` this is to prevent building lapack-lite as 64 bit integer mode which emscripten doesn't support;
3. patches modified (both are needed because the original files got changed and the original patches can no longer apply):
    1.  `fix-static-init-of-nditer-pywrap.patch`
    2. `use-local-blas-lapack.patch`

Notes: in numpy build file, I'm currently setting:
```
self.cpu_baseline = "avx"
self.cpu_dispatch = "avx"
```
I think this means there are still basic SIMD optimization built with this which is supported by emscripten. We might want to check if we want to enable it yet.

Issue: https://github.com/pyodide/pyodide/issues/1827

### Checklists

Not all of these steps are necessary for every PR.

- [ ] Add a [CHANGELOG](https://github.com/pyodide/pyodide/blob/main/docs/project/changelog.md) entry or explain why an entry is not needed
- [ ] Add / update tests
- [ ] Add new / update outdated documentation

### CODE REVIEW

File: packages/numpy/patches/disable_optimization.patch [hunk 1]
```diff
@@ -0,0 +1,13 @@
+diff --git a/numpy/distutils/command/build.py b/numpy/distutils/command/build.py
```
Comment (leopsidom) [inline packages/numpy/patches/disable_optimization.patch]: Unintentionally included this file. This patch is not needed.

**Comment by ryanking13** [comment]
This must've been a hard job. I appreciate your effort!

A minor comments:

1.

> * patches added:
>   
>   1. `fix-invalid-asm-instruction.patch` to fix invalid asm instructions
>   2. `limit-cpu-baseline.patch` this is to limit cpu baseline to `AVX` so it doesn't try to build SIMD feature AVX512 that emscripten doesn't support;
>   3. `not-build-lapack-lite-as-64-bit.patch` this is to prevent building lapack-lite as 64 bit integer mode which emscripten doesn't support;

I think adding these descriptions into the patch file would help other people who later see patch files.

2. It would be great if we add some tests for newly available features of numpy 1.21 (and if possible, 1.18, 1.19, 1.20 also). I'm not sure which are the core new functionalities of these numpy versions, but we could find out from their [release notes](https://numpy.org/devdocs/release.html).

**Comment by rth** [comment]
Thanks a lot for doing this work @leopsidom !

> I think this means there are still basic SIMD optimization built with this which is supported by emscripten. We might want to check if we want to enable it yet.

Right now the CI fails with,
```
emcc: error: Passing any of -msse, -msse2, -msse3, -mssse3, -msse4.1, -msse4.2, -mavx, -mfpu=neon flags also requires passing -msimd128!
```
which seems to be expected as discussed in https://github.com/emscripten-core/emscripten/issues/12714. I'm very in favor of investigating what we need to use SIMD a separate issue (and in particular check that it doesn't break Safari), for now maybe the easiest would be,
```
self.cpu_baseline = "min"
self.cpu_dispatch = "min"
``` 
to disable simd ?

>> Add / update tests
>
> It would be great if we add some tests for newly available features of numpy 1.21 (and if possible, 1.18, 1.19, 1.20 also). 

This is another subject we should improve on in Pyodide. Historically we have not run much tests for packages, because while have the ability to run the package test suites in pyodide https://github.com/pyodide/pyodide/issues/69 running them would be very slow though, and there are also some known failure (cf related open issues about numpy). So we have just written some sanity tests and that's it. However now we have much more CI resources than before and I think we could run them but maybe triggered manually and not for each commit.  Improving the package test situation is also somewhat a blocker for https://github.com/pyodide/pyodide/issues/1577

So there is a lot work that can be done there, for this PR for instance running the numpy test suite and report new issues would probably be sufficient. We can also do this after merging of this PR. The only thing maybe we should add a test for, is one of the operations that uses SIMD to make sure that calling it doesn't error after all the workaround that were done here.

**Comment by leopsidom** [comment]
> I think adding these descriptions into the patch file would help other people who later see patch files.

Good idea. I have added the description to the corresponding patch files.

> It would be great if we add some tests for newly available features of numpy 1.21

I think numpy has it's own test suite. Ideally we might want to see how we  can run the tests in pyodide as commented  by @rth.

**Comment by leopsidom** [comment]
> Right now the CI fails with,
>   `emcc: error: Passing any of -msse, -msse2, -msse3, -mssse3, -msse4.1, -msse4.2, -mavx, -mfpu=neon flags also requires passing -msimd128!`

Yeah, that means we have to pass an extra cflag `-msimd128` to emscripten when compiling SIMD features. It looks like an explicit restriction that emscripten tries to call out about what it actually supports.


> for now maybe the easiest would be
> self.cpu_baseline = "min"
> self.cpu_dispatch = "min"

Sounds good. I've switched both to `min`. Though I don't know if this completely disables SIMD. We might have to use `self.disable_optimization = True` to completely disable it.

**Comment by hoodmane** [comment]
Scipy build is failing because it can't link `npymath`.

**Comment by leopsidom** [comment]
Hmm yeah, will check a little later this week.

**Comment by leopsidom** [comment]
The `npymath` library issue can be fixed by updating `NUMPY_LIB` here: https://github.com/pyodide/pyodide/blob/main/packages/Makefile#L6. But it seems there are some more compatibility issue between numpy and scipy. 

> ModuleNotFoundError: No module named 'numpy.testing.decorators'

I think `numpy.testing.decorators`  has been removed in the newer version of numpy.

**Comment by leopsidom** [comment]
I think this will depend on success of this upgrade: https://github.com/pyodide/pyodide/pull/1293

**Comment by hoodmane** [comment]
Well it's still available as 
`from numpy.testing._private.decorators import setastest`
Maybe we can patch numpy to copy `numpy.testing._private.decorators` to `numpy.testing.decorators`.

**Comment by leopsidom** [comment]
Ahh, good point. Let me try that.

File: pyodide-build/pyodide_build/pywasmcross.py [hunk 2]
```diff
@@ -329,6 +329,8 @@ def handle_command(line, args, dryrun=False):
             continue
         if arg == "-lffi":
             continue
+        if arg == "-mpopcnt":
```
Comment (rth) [inline pyodide-build/pyodide_build/pywasmcross.py:332]: Maybe let's add a comment that this is a SSE 4.2 instruction, and we currently build without SIMD.

Just so it's easier to know what to re-enable once we want to build with SIMD.

File: packages/numpy/patches/add-back-decorators-module.patch [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
++function object need to use
++::
++
++  nose.tools.make_decorator(original_function)(decorator)
++
++in returning the decorator, in order to preserve meta-data such as
++function name, setup and teardown functions and so on - see
++``nose.tools`` for more information.
++
++"""
```
Comment (rth) [inline packages/numpy/patches/add-back-decorators-module.patch:24]: You could do 
```
from numpy.testing._private.decorators import *
```
in this patch maybe?

**Review by rth — COMMENTED** [review]
Thanks @leopsidom, great I see the CI passes!

A few minor comments below otherwise looks good. Could you also please merge upstream/main to sync? I merged https://github.com/pyodide/pyodide/pull/1708 recently which adds a few more patches to numpy, but you should be able to remove the `fix-random-double-fill.patch` with this update.

**Comment by leopsidom** [comment]
It seems the `fix-getters.patch` is also not needed: https://github.com/numpy/numpy/blob/main/numpy/core/src/multiarray/getset.c#L29 ?

**Comment by rth** [comment]
> It seems the fix-getters.patch is also not needed

It's possible, I wasn't 100% sure in what release it got included.

**Comment by rth** [comment]
No actually my initial comment was correct. That patch is [not included in 1.21.4](https://github.com/numpy/numpy/blob/c0b003e9c787ccab27f6fe57c154d7b881da5795/numpy/core/src/multiarray/getset.c#L29), so it's still necessary.

**Comment by leopsidom** [comment]
> That patch is not included in 1.21.4, so it's still necessary.

Yeah, I was looking at a newer commit @1.22.0.dev0

**Comment by leopsidom** [comment]
Also I'm seeing two patches not applied in `meta.yaml`: `0008-fix-dict.patch` and `make-int-return-values.patch`, wonder why we included them in the patches ?

File: pyodide-build/pyodide_build/pywasmcross.py (hunk 2 above)
Comment (leopsidom) [reply pyodide-build/pyodide_build/pywasmcross.py:333]: Added the comment -

**Comment by hoodmane** [comment]
> Also I'm seeing two patches not applied in meta.yaml: 0008-fix-dict.patch and make-int-return-values.patch, wonder why we included them in the patches ?

Maybe they aren't needed any more, but we just forgot to delete them?

**Comment by leopsidom** [comment]
Okay, makes sense. Let me just remove them -

**Comment by leopsidom** [comment]
Oh wait I think they are correct. The first one is Cython patch not a numpy patch. The 2nd does get included in meta.yaml.

File: packages/numpy/patches/add-back-decorators-module.patch (hunk 3 above)
Comment (leopsidom) [reply packages/numpy/patches/add-back-decorators-module.patch]: Great idea. I have updated the patch

**Comment by rth** [comment]
Thanks @leopsidom !

**Comment by hoodmane** [comment]
Thanks @leopsidom! This is very helpful. Though I think we should have a changelog entry for this?

**Comment by leopsidom** [comment]
No problems. Thanks for accepting the PR @rth. @hoodmane sure, I can take a look and add a changelog for it a little later.

**Comment by leopsidom** [comment]
> Thanks @leopsidom! This is very helpful. Though I think we should have a changelog entry for this?

Opened a [PR](https://github.com/pyodide/pyodide/pull/2054) to update the change log.

**Pull request merged**